src/CorePolicy.cc

```diff
@@ -0,0 +1,247 @@
+ */
+const double loadFactorThreshold = 1.0;
+
+/*
+ * Save the core fraction at which we ramped up based on load factor, so we
+ * can decide whether to ramp down.  Allocated in bootstrapLoadEstimator.
+ */
+double *utilizationThresholds = NULL;
+
+/*
+ * The difference in load, expressed as a fraction of a core, between a
+ * ramp-down threshold and the corresponding ramp-up threshold (i.e., we
+ * wait to ramp down until the load gets a bit below the point at 
+ * which we ramped up).
+ */
+const double idleCoreFractionHysteresis = 0.2;
+
+/*
+ * Do not ramp down if the percentage of occupied threadContext slots is
+ * above this threshold.
+ */
+const double SLOT_OCCUPANCY_THRESHOLD = 0.5;
+
+/*
+ * The period in ns over which we measure before deciding to reduce the number
+ * of cores we use.
+ */
+const uint64_t MEASUREMENT_PERIOD = 50 * 1000 * 1000;
+
+/* Is the core load estimator running? */
+bool loadEstimatorRunning = false;
+
+void coreLoadEstimator(CorePolicy* corePolicy);
+
+/*
+ * Choose a core to deschedule.  Return its coreId.
+ */
+int CorePolicy::chooseRemovableCore() {
+    CoreList* entry = threadClassCoreMap[defaultClass];
+    uint8_t minLoaded =
+        Arachne::occupiedAndCount[entry->map[0]]->load().numOccupied;
+    int minCoreId = entry->map[0];
+    for (uint32_t i = 1; i < threadClassCoreMap[defaultClass]->numFilled; i++) {
+        uint32_t coreId = entry->map[i];
+        if (Arachne::occupiedAndCount[coreId]->load().numOccupied < minLoaded) {
+            minLoaded = Arachne::occupiedAndCount[coreId]->load().numOccupied;
+            minCoreId = coreId;
+        }
+    }
+    return minCoreId;
+}
+
+/*
+ * Update threadClassCoreMap when a new core is added.  Callers must guarantee
+ * that this function is not called multiple times simultaneously, or at the
+ * same time as removeCore.
+ *
+ * This function will also bootstrap the coreLoadEstimator as soon as it has
+ * a core.
+ *
+ * \param coreId
+ *     The coreId of the new core.
+ */
+void CorePolicy::addCore(int coreId) {
+  CoreList* entry = threadClassCoreMap[defaultClass];
+  entry->map[entry->numFilled] = coreId;
+  entry->numFilled++;
+  if (!loadEstimatorRunning && !Arachne::disableLoadEstimation) {
+    loadEstimatorRunning = true;
+    runLoadEstimator();
+  }
+}
+
+/*
+ * Update threadClassCoreMap when a new core is removed.  Callers must
+ * guarantee that this function is not called multiple times simultaneously,
+ * or at the same time as addCore.
+ *
+ * \param coreId
+ *     The coreId of the doomed core.
+ */
+void CorePolicy::removeCore(int coreId) {
+  CoreList* entry = threadClassCoreMap[defaultClass];
+  entry->numFilled--;
+  int saveCoreId = entry->map[entry->numFilled];
+  for (uint32_t i = 0; i < entry->numFilled; i++) {
+    if (entry->map[i] == coreId) {
+      entry->map[i] = saveCoreId;
+      return;
+    }
+  }
+}
+
+/*
+ * Return the CoreList for a particular threadClass.  This function is not

```

Comment:
Again, the first sentence is at too low a level (e.g. what does the CoreList indicate?). Let's discuss.

Also, I think you need to be more clear about exactly what "stale data" can mean: for example, can it mean returning a null pointer?